src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,462 @@
+                                                  const AttributeValueEncoder::ListEncodeHelper & aEncoder)
+{
+    Delegate * delegate = GetDelegate(aPath.mEndpointId);
+    VerifyOrReturnError(IsValid(delegate) == CHIP_NO_ERROR, CHIP_ERROR_INCORRECT_STATE);
+
+    for (uint16_t i = 0; i < kMaxActionListLength; i++)
+    {
+        ActionStructStorage action;
+        CHIP_ERROR err = delegate->ReadActionAtIndex(i, action);
+
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(aEncoder.Encode(action));
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ActionsServer::ReadEndpointListAttribute(const ConcreteReadAttributePath & aPath,
+                                                    const AttributeValueEncoder::ListEncodeHelper & aEncoder)
+{
+    Delegate * delegate = GetDelegate(aPath.mEndpointId);
+    VerifyOrReturnError(IsValid(delegate) == CHIP_NO_ERROR, CHIP_ERROR_INCORRECT_STATE);
+
+    for (uint16_t i = 0; i < kMaxEndpointListLength; i++)
+    {
+        EndpointListStorage epList;
+
+        CHIP_ERROR err = delegate->ReadEndpointListAtIndex(i, epList);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(aEncoder.Encode(epList));
+    }
+    return CHIP_NO_ERROR;
+}
+
+bool ActionsServer::HaveActionWithId(EndpointId aEndpointId, uint16_t aActionId)
+{
+    Delegate * delegate = GetDelegate(aEndpointId);
+    VerifyOrReturnValue(IsValid(delegate) == CHIP_NO_ERROR, false);
+    return delegate->HaveActionWithId(aActionId);
+}
+
+template <typename RequestT, typename FuncT>
+void ActionsServer::HandleCommand(HandlerContext & handlerContext, FuncT func)
+{
+    if (!handlerContext.mCommandHandled && (handlerContext.mRequestPath.mCommandId == RequestT::GetCommandId()))
+    {
+        RequestT requestPayload;
+
+        // If the command matches what the caller is looking for, let's mark this as being handled
+        // even if errors happen after this. This ensures that we don't execute any fall-back strategies
+        // to handle this command since at this point, the caller is taking responsibility for handling
+        // the command in its entirety, warts and all.
+        //
+        handlerContext.SetCommandHandled();
+
+        if (DataModel::Decode(handlerContext.mPayload, requestPayload) != CHIP_NO_ERROR)
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath,
+                                                     Protocols::InteractionModel::Status::InvalidCommand);
+            return;
+        }
+
+        if (HaveActionWithId(handlerContext.mRequestPath.mEndpointId, requestPayload.actionID))
+        {
+            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath, Protocols::InteractionModel::Status::NotFound);
+            return;
```

Comment:
This code looks wrong.  Was this tested?

Reply:
Thanks for pointing out. It is tested and working as expected.